### PR TITLE
Pipeline.

### PR DESCRIPTION
**Pull request by ashleypittman**
Squash and rebase of PR-1

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                            additionalBuildArgs "-t ${sanitized_JOB_NAME}-centos7 " + '$BUILDARGS'
+                        }
+                    }
+                    steps {
+                        sconsBuild clean: "_build.external${arch}"
+                        stash name: 'CentOS-install', includes: 'install/**'
+                        stash name: 'CentOS-build-vars', includes: ".build_vars${arch}.*"
+                    }
+                    post {
+                        always {
```
Comment (brianjmurrell) [inline Jenkinsfile]: You need to specify a _lightweight_ node for this step so that it does not tie up more useful agents.

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                            recordIssues enabledForFailure: true,
+                                         aggregatingResults: true,
+                                         id: "analysis-centos7",
+                                         tools: [ clang(), cppCheck() ],
```
Comment (brianjmurrell) [inline Jenkinsfile]: `clang()`?  Does IOF build with _clang_ by default or gcc, like cart and daos do?

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                            dir 'utils/docker'
+                            label 'docker_runner'
+                            additionalBuildArgs "-t ${sanitized_JOB_NAME}-ubuntu18.04 " + '$BUILDARGS'
+                        }
+                    }
+                    steps {
+                        sconsBuild clean: "_build.external${arch}"
+                    }
+                    post {
+                        always {
```
Comment (brianjmurrell) [inline Jenkinsfile]: Need the _lightweight_ agent here also.

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                            recordIssues enabledForFailure: true,
+                                         aggregatingResults: true,
+                                         id: "analysis-ubuntu18",
+                                         tools: [ clang(), cppCheck() ],
```
Comment (brianjmurrell) [inline Jenkinsfile]: Same question about _clang_.

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                                         filters: [excludeFile('.*\\/_build\\.external\\/.*'),
```
Comment (brianjmurrell) [inline Jenkinsfile]:
```suggestion
                                         filters: [excludeFile('.*\\/_build\\.external-Linux\\/.*'),
```

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                                                   excludeFile('_build\\.external\\/.*')]
```
Comment (brianjmurrell) [inline Jenkinsfile]:
```suggestion
                                                   excludeFile('_build\\.external-Linux\\/.*')]
```

File: Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+                    post {
+                        always {
+                            recordIssues enabledForFailure: true,
+                                         aggregatingResults: true,
+                                         id: "analysis-centos7",
+                                         tools: [ clang(), cppCheck() ],
+                                         filters: [excludeFile('.*\\/_build\\.external-Linux\\/.*'),
+                                                   excludeFile('_build\\.external-Linux\\/.*')]
+                        }
+                    }
```
Comment (brianjmurrell) [inline Jenkinsfile]: You need a `success { }` block here that removes `_build.external${arch}` for space management on Jenkins.  See the cart `Jenkinsfile`.

File: Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                    post {
+                        always {
+                            recordIssues enabledForFailure: true,
+                                         aggregatingResults: true,
+                                         id: "analysis-ubuntu18",
+                                         tools: [ clang(), cppCheck() ],
+                                         filters: [excludeFile('.*\\/_build\\.external\\/.*'),
+                                                   excludeFile('_build\\.external\\/.*')]
+                        }
```
Comment (brianjmurrell) [inline Jenkinsfile]: Ditto for the `success { }` block.

**Comment by ashleypittman** [comment]
> You are not interested in a more fuller array of build distros/compilers on patch landings (i.e. to master) or a daily build of master like cart does?

I'm interested in getting something working and landed first and foremost.

More compilers would be good, enabling the unit, valgrind and fault-injection tests come next, as does setting up some weekly longer-duration testing.

File: Jenkinsfile [hunk 9]
```diff
@@ -102,13 +102,19 @@ pipeline {
...
-                                         aggregatingResults: true,
-                                         id: "analysis-centos7",
-                                         tools: [ clang(), cppCheck() ],
-                                         filters: [excludeFile('.*\\/_build\\.external-Linux\\/.*'),
-                                                   excludeFile('_build\\.external-Linux\\/.*')]
+                            node('lightweight') {
+                                recordIssues enabledForFailure: true,
+                                             aggregatingResults: true,
+                                             id: "analysis-centos7",
+                                             tools: [ cppCheck() ],
```
Comment (brianjmurrell) [inline Jenkinsfile]: I think you still want to specify a compiler here, like `gcc4()` if the build is using _gcc_.  That being missing might be what's causing the:

```
ID cppcheck is already used by another action: io.jenkins.plugins.analysis.core.model.ResultAction for CPPCheck
```

error that you were getting.  Just a guess mind you.

File: Jenkinsfile [hunk 10]
```diff
@@ -102,13 +102,19 @@ pipeline {
...
+                                             filters: [excludeFile('.*\\/_build\\.external-Linux\\/.*'),
+                                                       excludeFile('_build\\.external-Linux\\/.*')]
+			    }
                         }
+                        success {
+                            sh "rm -rf _build.external${arch}"
+                        }
+
```
Comment (brianjmurrell) [inline Jenkinsfile]: Remove this empty line?

File: Jenkinsfile [hunk 11]
```diff
@@ -125,13 +131,19 @@ pipeline {
...
-                                         aggregatingResults: true,
-                                         id: "analysis-ubuntu18",
-                                         tools: [ clang(), cppCheck() ],
-                                         filters: [excludeFile('.*\\/_build\\.external\\/.*'),
-                                                   excludeFile('_build\\.external\\/.*')]
+			    node('lightweight') {
+                                recordIssues enabledForFailure: true,
+                                             aggregatingResults: true,
+                                             id: "analysis-ubuntu18",
+                                             tools: [ cppCheck() ],
```
Comment (brianjmurrell) [inline Jenkinsfile]: Ditto for the compiler.

File: Jenkinsfile [hunk 12]
```diff
@@ -125,13 +131,19 @@ pipeline {
...
+                                             filters: [excludeFile('.*\\/_build\\.external-Linux\\/.*'),
+                                                       excludeFile('_build\\.external-Linux\\/.*')]
+			    }
                         }
+                        success {
+                            sh "rm -rf _build.external${arch}"
+                        }
+
```
Comment (brianjmurrell) [inline Jenkinsfile]: Ditto for the empty line?

**Comment by ashleypittman** [comment]
> Minor nits that can be handled in a future patches as they need to be done for Cart/Daos also.

I'm not trying to make this perfect, or fix things that are present in Cart/Daos, just get IOF out the gate.

> The credential for checkPatch can be passed as credential instead of set as a global environment variable.

I can look at this but I'm not sure how to do it, so I'll leave it as it is for now.

> We plan to rename the dockerfile names to not have colons in them. But we are also looking at moving the dockerfiles to scons_local.

I've had to customise the Docker files for iof, compared to CaRT, mostly adding new packages I think.

**Pull request merged**